### PR TITLE
Fix sparse benchmark

### PR DESCRIPTION
**Pull request by kalyc**
### Summary
Keras-MXNet benchmark script resulted in segmentation fault error initially. Added check to ensure that it doesn't happen. 

`Note` - 
Benchmark script currently works well only with batch size <= (10) train_data.shape[1] size

For dimensions > 10 this is the error we get - 
`
File "benchmark/sparse/linear_regression/run_sparse_benchmark.py", line 40, in <module>
    invoke_benchmark(args.batch, args.epochs)
  File "benchmark/sparse/linear_regression/run_sparse_benchmark.py", line 25, in invoke_benchmark
    epochs=int(epochs))
  File "/Users/kalyanee/Documents/kalyc-keras-1/keras-apache-mxnet/benchmark/sparse/linear_regression/keras_sparse_model.py", line 26, in run_benchmark
    batch_size=batch_size)
  File "/anaconda2/envs/mxnet/lib/python3.4/site-packages/keras_mxnet-2.2.2-py3.4.egg/keras/engine/training.py", line 1037, in fit
  File "/anaconda2/envs/mxnet/lib/python3.4/site-packages/keras_mxnet-2.2.2-py3.4.egg/keras/engine/training_arrays.py", line 199, in fit_loop
  File "/anaconda2/envs/mxnet/lib/python3.4/site-packages/keras_mxnet-2.2.2-py3.4.egg/keras/backend/mxnet_backend.py", line 4982, in train_function
  File "/anaconda2/envs/mxnet/lib/python3.4/site-packages/keras_mxnet-2.2.2-py3.4.egg/keras/backend/mxnet_backend.py", line 4942, in _adjust_module
  File "/anaconda2/envs/mxnet/lib/python3.4/site-packages/mxnet/module/module.py", line 471, in reshape
    self._exec_group.reshape(self._data_shapes, self._label_shapes)
  File "/anaconda2/envs/mxnet/lib/python3.4/site-packages/mxnet/module/executor_group.py", line 382, in reshape
    self.bind_exec(data_shapes, label_shapes, reshape=True)
  File "/anaconda2/envs/mxnet/lib/python3.4/site-packages/mxnet/module/executor_group.py", line 358, in bind_exec
    allow_up_sizing=True, **dict(data_shapes_i + label_shapes_i))
  File "/anaconda2/envs/mxnet/lib/python3.4/site-packages/mxnet/executor.py", line 423, in reshape
    new_arg_dict[name] = arr.reshape(new_shape)
  File "/anaconda2/envs/mxnet/lib/python3.4/site-packages/mxnet/ndarray/sparse.py", line 157, in reshape
    raise NotSupportedForSparseNDArray(self.reshape, None, shape)
mxnet.base.NotSupportedForSparseNDArray: Function reshape with arguments (<class 'tuple'>) is not supported for SparseNDArray and only available in NDArray.
`
### PR Overview

- [n] This PR requires new unit tests [y/n] (make sure tests are included)
- [n] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [n] This PR is backwards compatible [y/n]
- [n] This PR changes the current API [y/n]

### Solution
Added workaround for the reshape error by slicing data before training

### CODE REVIEW

File: benchmark/sparse/linear_regression/keras_sparse_model.py [hunk 1]
```diff
@@ -7,12 +7,14 @@
 from keras import Model
 from keras.layers import Dense, Input
 from keras.optimizers import SGD
+from keras import backend as K
 
 
 def run_benchmark(train_data, train_label, eval_data, eval_label, batch_size, epochs):
-    inputs = Input(shape=(10,), sparse=True)
+    inputs = Input(shape=(10,), dtype='float32', sparse=True)
```
Comment (roywei) [inline benchmark/sparse/linear_regression/keras_sparse_model.py]: why we specify dtype here?

File: benchmark/sparse/linear_regression/keras_sparse_model.py [hunk 2]
```diff
@@ -7,12 +7,14 @@
...
     predictions = Dense(10, activation='linear')(inputs)
-    model = Model(input=inputs, output=predictions)
+    model = Model(outputs=predictions, inputs=inputs)
```
Comment (roywei) [inline benchmark/sparse/linear_regression/keras_sparse_model.py]: keep it the original order for readability.

File: benchmark/sparse/linear_regression/keras_sparse_model.py [hunk 3]
```diff
@@ -23,7 +25,8 @@ def run_benchmark(train_data, train_label, eval_data, eval_label, batch_size, ep
               epochs=epochs,
               batch_size=batch_size)
 
-    print("Keras-TF Sparse Benchmark Results")
+    print("Keras Sparse Benchmark Results")
+    print("Backend: ", K.backend())
     print("Batch Size")
```
Comment (roywei) [inline benchmark/sparse/linear_regression/keras_sparse_model.py]: Keep logging style consistent for backend, batch size, etc. print("Name: ", value)

File: benchmark/sparse/linear_regression/keras_sparse_model.py (hunk 2 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/keras_sparse_model.py]: This change is taking into account the new Keras-2 api for Model

Comment (roywei) [reply benchmark/sparse/linear_regression/keras_sparse_model.py]: I mean to put inputs arguments before outputs argument, they should be equivalent right? https://keras.io/models/model/

Comment (kalyc) [reply benchmark/sparse/linear_regression/keras_sparse_model.py]: Oh the ordering, yea I can change that.

File: benchmark/sparse/linear_regression/keras_sparse_model.py (hunk 3 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/keras_sparse_model.py]: fixed

File: benchmark/sparse/linear_regression/mxnet_sparse_model.py [hunk 4]
```diff
@@ -6,35 +6,50 @@
 import mxnet as mx
```
Comment (roywei) [inline benchmark/sparse/linear_regression/mxnet_sparse_model.py]: add a reference to mxnet sparse linear regression example.
In case of future changes so it's easy to refer and maintain

File: benchmark/sparse/linear_regression/mxnet_sparse_model.py [hunk 5]
```diff
@@ -6,35 +6,50 @@
...
+
+    eval_iter = mx.io.NDArrayIter(eval_data, eval_label, batch_size=batch_size,
+                                  label_name='label', last_batch_handle='discard')
 
     X = mx.sym.Variable('data', stype='csr')
-    Y = mx.symbol.Variable('lin_reg_label', stype='csr')
-    fully_connected_layer = mx.sym.FullyConnected(data=X, name='fc1', num_hidden=10)
-    lro = mx.sym.LinearRegressionOutput(data=fully_connected_layer, label=Y, name="lro")
+    weight = mx.symbol.Variable('weight', stype='row_sparse', shape=(train_data.shape[1], 1))
+    pred = mx.sym.sparse.dot(X, weight)
```
Comment (roywei) [inline benchmark/sparse/linear_regression/mxnet_sparse_model.py]: please add bias here according to example:
https://mxnet.incubator.apache.org/tutorials/sparse/train.html

in keras, when you add Dense layer, bias is added by default

File: benchmark/sparse/linear_regression/run_sparse_benchmark.py [hunk 6]
```diff
@@ -3,38 +3,60 @@
...
-                                 eval_label=sparse.csr_matrix(eval_label.asnumpy()),
-                                 batch_size=int(batch_size),
-                                 epochs=int(epochs))
 
-    print("Running MXNet sparse benchmark script")
-    mxnet_sparse_model.run_benchmark(train_data=train_data, train_label=train_label, eval_data=eval_data,
-                                     eval_label=eval_label, batch_size=int(batch_size), epochs=int(epochs))
+def invoke_benchmark(batch_size, epochs):
+    # Fix seed for randomness
+    np.random.rand(7)
```
Comment (roywei) [inline benchmark/sparse/linear_regression/run_sparse_benchmark.py]: can we remove this and run a few time (e.g. 5) and take an average of the speed.

File: benchmark/sparse/linear_regression/results.md [hunk 7]
```diff
@@ -9,22 +9,24 @@
...
-|-----|-----|-----|----- |-----|
-|  C5.18X Large |   0  | 16  | 51.15 sec | 71.85 sec |
-|  C5.18X Large |   0  | 32 | 20.82 sec | 54.31 sec |
-|  C5.18X Large |   0  | 64  | 13.13 sec | 19.20 sec |
-|  C5.18X Large |   0  | 128  | 5.72 sec | 9.86 sec |
+##### Sparse data
+###### Using 25 epochs
+| Instance Type | GPUs  | Batch Size  | MXNet (Time/Epoch) | Keras-MXNet (Time/Epoch) | Keras-TensorFlow (Time/Epoch)  |
+|-----|-----|-----|-----|-----|-----|
+| C5.18XLarge |   0  | 16  | 52.89 sec | 334.32 sec | 237.28 sec |
```
Comment (roywei) [inline benchmark/sparse/linear_regression/results.md]: Can we report an average of a few runs, not a fixed seed result

File: keras/utils/data_utils.py [hunk 8]
```diff
@@ -299,6 +300,11 @@ def validate_file(fpath, file_hash, algorithm='auto', chunk_size=65535):
         return False
 
 
+def prepare_sliced_sparse_data(data, batch_size):
```
Comment (roywei) [inline keras/utils/data_utils.py]: a unit test for this function? what if `data.shape[0] < batch_size` or `data.shape[0] is None`? maybe a warning before returning an empty list

File: benchmark/sparse/linear_regression/mxnet_sparse_model.py (hunk 5 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/mxnet_sparse_model.py]: Bias is added in Keras only if present. We dont need to add bias explicitly. As I am not adding bias in the MXNet model, its not added in the keras model either - https://github.com/awslabs/keras-apache-mxnet/blob/master/keras/layers/core.py#L875

File: benchmark/sparse/linear_regression/mxnet_sparse_model.py (hunk 4 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/mxnet_sparse_model.py]: done

File: benchmark/sparse/linear_regression/mxnet_sparse_model.py (hunk 5 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/mxnet_sparse_model.py]: added

File: benchmark/sparse/linear_regression/results.md (hunk 7 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/results.md]: done

File: keras/utils/data_utils.py (hunk 8 above)
Comment (kalyc) [reply keras/utils/data_utils.py]: added

File: tests/keras/utils/data_utils_test.py [hunk 9]
```diff
@@ -368,5 +370,33 @@ def test_finite_generator_enqueuer_processes():
     enqueuer.stop()
 
 
+def test_prepare_sparse_sliced_data():
+    test_train_data = mx.test_utils.rand_ndarray((10, 10), 'csr', 0.01)
```
Comment (roywei) [inline tests/keras/utils/data_utils_test.py]: use numpy to create random data of (10,10) and remove import mxnet, this is a general until function, should be tested with numpy and should not rely on mxnet

File: tests/keras/utils/data_utils_test.py [hunk 10]
```diff
@@ -368,5 +370,33 @@ def test_finite_generator_enqueuer_processes():
...
+    test_train_data = None
+    batch_size = 3
+
+    with pytest.warns(UserWarning):  # Warning is thrown when data is None
+        result = prepare_sliced_sparse_data(test_train_data, batch_size)
+        assert result is None
+
+
+def test_prepare_sparse_sliced_data_incorrect_dimensions():
+    test_train_data = mx.test_utils.rand_ndarray((2, 2), 'csr', 0.01)
```
Comment (roywei) [inline tests/keras/utils/data_utils_test.py]: same as above

File: tests/keras/utils/data_utils_test.py [hunk 11]
```diff
@@ -16,9 +16,11 @@
 from keras.utils import GeneratorEnqueuer
 from keras.utils import OrderedEnqueuer
 from keras.utils import Sequence
+import mxnet as mx
```
Comment (roywei) [inline tests/keras/utils/data_utils_test.py]: use numpy instead of mxnet

File: tests/keras/utils/data_utils_test.py (hunk 9 above)
Comment (kalyc) [reply tests/keras/utils/data_utils_test.py]: fixed

File: tests/keras/utils/data_utils_test.py (hunk 10 above)
Comment (kalyc) [reply tests/keras/utils/data_utils_test.py]: fixed

File: tests/keras/utils/data_utils_test.py (hunk 11 above)
Comment (kalyc) [reply tests/keras/utils/data_utils_test.py]: fixed
